### DISCUSSION TITLE
[Question] How to wait for a dialog to return a value?

### DISCUSSION THREAD

**Discussion by kojix2** [hugopl/gtk4.cr#40] [Q&A] [answered]
Hi there!

Thank you for creating a practical GTK4 binding for the Crystal language.

I am working on a tool that shows data in a tabular format like this. 
![image](https://user-images.githubusercontent.com/5798442/193750563-04f17bcb-e4aa-4c49-85ed-a9578a0faec5.png)

To add the ability to open a file, I wrote the following code.

```crystal
      def get_file_path
        file_path = ""
        dialog = Gtk::FileChooserDialog.new(
          application: @app,
          title: "Open File",
          action: Gtk::FileChooserAction::Open,
          transient_for: window,
          modal: true
        )
        dialog.add_button("Cancel", Gtk::ResponseType::Cancel.value)
        dialog.add_button("Open", Gtk::ResponseType::Accept.value)
        dialog.response_signal.connect do |response|
          case Gtk::ResponseType.from_value(response)
          when .cancel?
          when .accept?
            file_path = dialog.file.try(&.path)
          end
          dialog.destroy
        end
        dialog.present
        file_path
      end
```

Here I tried to get the file path. This works. 
The problem is that the method always return `""`. 
It doesn't wait for the dialog to exit.

What am I missing?

**Comment by hugopl** [comment, marked as answer]
GTK is event based, the `get_file_path` method you wrote will return before the file chooser gets closed, so you need to handle what happens when it closes in the response signal instead, something like call a method `file_path_changed(Path)` from there, etc... IIRC GTK3 had a version that blocked the execution by creating another event loop internally, this is also possible on GTK4 but if I'm not wrong it requires you to create this new event loop by yourself.

So my suggestion is to embrace the asynchronous behavior of the API.

**Reply by kojix2** [reply]
Thank you very much.
So instead of waiting, I should just connect to the method that handles the signal when it occurs?
I hadn't realized that, but you are right.
